plugin/trino-sqlserver/src/test/java/io/trino/plugin/sqlserver/TestSqlServerTypeMapping.java

```diff
@@ -66,6 +95,127 @@ public void testVarbinary()
+                .addRoundTrip("TIMESTAMP '1970-01-01 00:00:00.123000'", "TIMESTAMP '1970-01-01 00:00:00.123000'")
+                .addRoundTrip("TIMESTAMP '1970-01-01 00:00:00.999'", "TIMESTAMP '1970-01-01 00:00:00.999'")
+                // max supported precision
+                .addRoundTrip("TIMESTAMP '1970-01-01 00:00:00.1234567'", "TIMESTAMP '1970-01-01 00:00:00.1234567'")
+
+                .addRoundTrip("TIMESTAMP '2020-09-27 12:34:56.1'", "TIMESTAMP '2020-09-27 12:34:56.1'")
+                .addRoundTrip("TIMESTAMP '2020-09-27 12:34:56.9'", "TIMESTAMP '2020-09-27 12:34:56.9'")
+                .addRoundTrip("TIMESTAMP '2020-09-27 12:34:56.123'", "TIMESTAMP '2020-09-27 12:34:56.123'")
+                .addRoundTrip("TIMESTAMP '2020-09-27 12:34:56.123000'", "TIMESTAMP '2020-09-27 12:34:56.123000'")
+                .addRoundTrip("TIMESTAMP '2020-09-27 12:34:56.999'", "TIMESTAMP '2020-09-27 12:34:56.999'")
+                // max supported precision
+                .addRoundTrip("TIMESTAMP '2020-09-27 12:34:56.1234567'", "TIMESTAMP '2020-09-27 12:34:56.1234567'")
+
+                // round down
+                .addRoundTrip("TIMESTAMP '1970-01-01 00:00:00.12345671'", "TIMESTAMP '1970-01-01 00:00:00.1234567'")
+
+                // nanoc round up, end result rounds down
+                .addRoundTrip("TIMESTAMP '1970-01-01 00:00:00.1234567499'", "TIMESTAMP '1970-01-01 00:00:00.1234567'")
+                .addRoundTrip("TIMESTAMP '1970-01-01 00:00:00.123456749999'", "TIMESTAMP '1970-01-01 00:00:00.1234567'")
+
+                // round up
+                .addRoundTrip("TIMESTAMP '1970-01-01 00:00:00.12345675'", "TIMESTAMP '1970-01-01 00:00:00.1234568'")
+
+                // max precision
+                .addRoundTrip("TIMESTAMP '1970-01-01 00:00:00.111222333444'", "TIMESTAMP '1970-01-01 00:00:00.1112223'")
+
+                // round up to next second
+                .addRoundTrip("TIMESTAMP '1970-01-01 00:00:00.99999995'", "TIMESTAMP '1970-01-01 00:00:01.0000000'")
+
+                // round up to next day
+                .addRoundTrip("TIMESTAMP '1970-01-01 23:59:59.99999995'", "TIMESTAMP '1970-01-02 00:00:00.0000000'")
+
+                // negative epoch
+                .addRoundTrip("TIMESTAMP '1969-12-31 23:59:59.99999995'", "TIMESTAMP '1970-01-01 00:00:00.0000000'")
+                .addRoundTrip("TIMESTAMP '1969-12-31 23:59:59.999999949999'", "TIMESTAMP '1969-12-31 23:59:59.9999999'")
+                .addRoundTrip("TIMESTAMP '1969-12-31 23:59:59.99999994'", "TIMESTAMP '1969-12-31 23:59:59.9999999'")
+
+                // CTAS with Presto, where the coercion is done by the connector
+                .execute(getQueryRunner(), trinoCreateAsSelect(getSession(), "test_timestamp_coercion_connector"))
+                // INSERT with Presto, where the coercion is done by the engine
+                .execute(getQueryRunner(), trinoCreateAndInsert(getSession(), "test_timestamp_coercion_engine"));
```

Comment:
This tests roundtrip, but ideally we'd test with `CreateAndInsertDataSetup` via SQL Server too, so we make sure we don't have read/write bugs masking each other

Reply:
I considered that but these tests generate queries with `TIMESTAMP` and SQL Server expects `DATETIME2`. I would have to create a copy of `testTimestampCoercion()`. Is this ok?

Reply:
Also, SQL Server doesn't support a regular CTAS query, it's using `SELECT INTO`:

* https://docs.microsoft.com/en-us/sql/t-sql/queries/select-into-clause-transact-sql?view=sql-server-ver15
* https://github.com/trinodb/trino/blob/master/plugin/trino-sqlserver/src/main/java/io/trino/plugin/sqlserver/SqlServerClient.java#L173

Reply:
This would also require overloading `TestTable` that actually contains `CREATE TABLE AS` in the constructor. It would require copying quite a lot of code because of so many private properties. I don't feel very confident about doing this.

Reply:
when all types are specified `CreateAndInsertDataSetup`   will use ordinary CT + Insert

https://github.com/trinodb/trino/blob/03d8e5abc686c5c4a9f96ca14db07e2aed880174/testing/trino-testing/src/main/java/io/trino/testing/datatype/CreateAndInsertDataSetup.java#L77

>  I would have to create a copy of testTimestampCoercion(). Is this ok?

totally; or within same method

Reply:
Figured it out, it was easier than I expected, thanks! Closing this discussion as the whole commit requires another review.